e2e-tests/src/tests/constraints.test.ts

```diff
@@ -0,0 +1,129 @@
+    };
+    plan_id = await req.createPlan(request, plan_input);
+
+    expect(plan_id).not.toBeNull();
+    expect(plan_id).toBeDefined();
+    expect(typeof plan_id).toEqual("number");
+  });
+
+  test("Add Activity to Plan", async ({ request }) => {
+    const activityToInsert : ActivityInsertInput = {
+      arguments : {
+        "biteSize": 1
+      },
+      plan_id: plan_id,
+      type : "BiteBanana",
+      start_offset : activity_offset_hours + "h",
+    };
+    const activity_id = await req.insertActivity(request, activityToInsert);
+
+    expect(activity_id).not.toBeNull();
+    expect(activity_id).toBeDefined();
+  });
+
+});
+
+test.describe("Constraints", () => {
+  test("Add Constraint to Plan", async ({ request }) => {
+    const constraint: ConstraintInsertInput = {
+      name: constraint_name,
+      definition: "export default (): Constraint => Real.Resource(\"/fruit\").equal(Real.Resource(\"/peel\"))",
+      description: "",
+      model_id: null,
+      plan_id,
+    };
+    constraint_id = await req.insertConstraint(request, constraint);
+
+    expect(constraint_id).not.toBeNull();
+    expect(constraint_id).toBeDefined();
+  });
+
+  test("Check there are no violations yet", async ({ request }) => {
+    const violations: ConstraintViolation[] = await req.checkConstraints(request, plan_id);
+
+    expect(violations).not.toBeNull();
+    expect(violations).toBeDefined();
+    expect(violations).toHaveLength(0);
+  });
+
+  test("Run simulation", async ({ request }) => {
+    const resp: SimulationResponse = await awaitSimulation(request, plan_id);
+
+    expect(resp.status).toEqual("complete");
+  });
+
+  test("Check there is one violation", async ({ request }) => {
+    const violations: ConstraintViolation[] = await req.checkConstraints(request, plan_id);
+
+    expect(violations).not.toBeNull();
+    expect(violations).toBeDefined();
+    expect(violations).toHaveLength(1);
+
+    violation = violations[0];
+
+    expect(violation).not.toBeNull();
+    expect(violation).toBeDefined();
+  });
+
+  test("Check the violation is the expected one", async () => {
+    expect(violation.constraintName).toEqual(constraint_name)
+    expect(violation.constraintId).toEqual(constraint_id)
+    expect(violation.associations.resourceIds).toHaveLength(2);
+    expect(violation.associations.resourceIds).toContain("/fruit");
+    expect(violation.associations.resourceIds).toContain("/peel");
+  });
+
+  test("Check violation starts and ends as expected", async () => {
+    const plan_start_unix = time.getUnixEpochTime(plan_start_timestamp);
+    const plan_end_unix = time.getUnixEpochTime(plan_end_timestamp);
+    const plan_duration_micro = (plan_end_unix - plan_start_unix) * 1000;
+    const activity_offset_micro = activity_offset_hours * 60 * 60 * 1000 * 1000;
+
+    expect(violation.windows[0].start).toEqual(activity_offset_micro);
+    expect(violation.windows[0].end).toEqual(plan_duration_micro);
+  });
```

Comment:
These tests are missing a "Teardown" step to clean the DB of any testing artifacts (ie the things inserted during "Initialize"

Reply:
Good call, I've added a "Clean Up" test

Reply:
Less important, but another e2e test that may be useful is that there are no constraint violations after fixing the violation and resimulating.

Reply:
I've added this additional test